src/main/java/frc/robot/subsystems/Swerve.java

```diff
@@ -0,0 +1,155 @@
+package frc.robot.subsystems;
+
+import com.ctre.phoenix6.hardware.Pigeon2;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.math.kinematics.*;
+import edu.wpi.first.units.Measure;
+import edu.wpi.first.units.Units;
+import edu.wpi.first.units.Voltage;
+import edu.wpi.first.wpilibj.RobotController;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj.sysid.SysIdRoutineLog;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import edu.wpi.first.wpilibj2.command.sysid.SysIdRoutine;
+import frc.robot.RobotMap;
+
+import static edu.wpi.first.units.Units.Meters;
+
+public class Swerve extends SubsystemBase {
+
+
+    private final Pigeon2 pigeon;
+    private final SwerveDriveOdometry odometry;
+    private final SwerveModule[] swerveModules;
+    private final SwerveDriveKinematics kinematics;
+    SysIdRoutine sysIdRoutine;
+    public Swerve(SwerveModule[] swerveModules){
+        sysIdRoutine = new SysIdRoutine(new SysIdRoutine.Config(), new SysIdRoutine.Mechanism(this::volatageDrive,this::sysidLog,this));
+        this.swerveModules = swerveModules;
+        double distance = RobotMap.DISTANCE_MID_TO_CON;
+        this.kinematics = new SwerveDriveKinematics(
+                new Translation2d(distance, distance),
+                new Translation2d(distance, -distance),
+                new Translation2d(-distance, distance),
+                new Translation2d(-distance,-distance)
+        );
+        pigeon = new Pigeon2(RobotMap.PIGEON);
+        odometry = new SwerveDriveOdometry(kinematics,pigeon.getRotation2d(),getModulesPosition());
+
+    }
+
+    public Pose2d getPose() {
+        return odometry.getPoseMeters();
+    }
+
+    public SwerveModulePosition[] getModulesPosition(){
+        SwerveModulePosition[] swerveModulePositions = new SwerveModulePosition[4];
+        for(int i =0; i<4; i++){
+            swerveModulePositions[i] = swerveModules[i].getModulePosition();
+        }
+
+        return swerveModulePositions;
+    }
+
+    public SwerveModuleState[] getModuleStates(){
+        SwerveModuleState[] swerveModuleStates = new SwerveModuleState[4];
+        for(int i =0; i <4; i++){
+            swerveModuleStates[i] = swerveModules[i].getModuleStates();
+        }
+
+        return swerveModuleStates;
+    }
+
+    public ChassisSpeeds getSpeeds(){
+        return kinematics.toChassisSpeeds(new SwerveDriveKinematics.SwerveDriveWheelStates(getModuleStates()));
+    }
+
+    private void sysidLog(SysIdRoutineLog log) {
```

Comment:
for class organization: you should put private methods _after_ all the public methods